verify npm package

le-cong opened the pull request:
Fixes #75

le-cong commented on .github/workflows/ci.yml:
```diff
@@ -30,6 +30,8 @@ jobs:
         run: npm run ci:lint
       - name: Run Tests
         run: npm run ci:test
+        env:
```
needed to pass tests which access npm

le-cong commented:
introducing a new step in `publish-beta.yml` (between `Publish Beta` and `Comment npm publish result`), e.g. [smartdata test](https://github.com/checkdigit/smartdata/blob/199173c29644a6dd02ea3efa04364b25d58d5d57/.github/workflows/publish-beta.yml#L49)

carlansley commented on src/validate-npm-package/validate-npm-package.ts:
```diff
@@ -0,0 +1,95 @@
+async function obtainNpmPackage(packageVersion: string, workFolder: string): Promise<void> {
+  await addNPMRCFile(workFolder);
+
+  const fullCommandLine = `npm pack ${packageVersion}`;
```
could we do an `npm install` somewhere instead?  then we could do an `import()` and make sure all the machinery works (including the package.json is ok)

carlansley commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@checkdigit/github-actions",
-  "version": "2.0.0",
+  "version": "2.0.1",
```
since we're adding a new action, think this should be a MINOR

carlansley commented on src/validate-npm-package/validate-npm-package.ts:
```diff
@@ -0,0 +1,95 @@
+async function verifyImportEntryPoints(packageFolder: string): Promise<void> {
+  const packageJson = JSON.parse(await fs.readFile(`${packageFolder}/package.json`, 'utf8')) as PackageJson;
+  const isEsm = packageJson.type === 'module';
+  const bundleFolder = isEsm ? 'dist-mjs' : 'dist';
```
if `npm install` isn't an option, think it would be better to simulate what node does, i.e. read the package.json and figure out the entry point from that.

adcreare replied:
I think npm install and then doing what deployer does (import and then ping operation) would be good.

carlansley replied:
this is for any package (not necessarily a service)?  I was thinking an npm install + importing it from some generated typescript.  then compile & run.  that way you know the types are ok and the entry point is correct.

carlansley replied:
in the new-new world a service is started by `await`ing the default export of the main entry point.  so this would expose an issue with API/service code once that starts rolling out.  also, I feel the primary job of this action is to validate the npm package itself, which is effectively whatever is exported from the root index.ts, vs functionality as a service.  we have other things that do that.

le-cong replied:
that makes total sense

le-cong replied:
this is probably a known limitation using `import`: if the package references excluded(e.g. from *.test.ts) types or values but doesn't export them through its root `index.ts`, then no error will be reported.
on the other hand, using `node` cli to load that package's `index.ts` can expose the problem though.

le-cong commented:
as part of the latest commit, i've updated tests quite a bit mainly because:

- the `check-label` tests are tangled with the current version, which makes the test tricky and flaky;
- fixed other flaky tests dealing with temporary folders
- refactored github nock to make the usage of different PRs easier to understand

adcreare reviewed (approved):
LGTM - very comprehensive

carlansley commented on src/comment-npm-publish/index.ts:
```diff
@@ -18,16 +18,4 @@ export async function main(): Promise<void> {
-    // eslint-disable-next-line unicorn/no-process-exit
-    process.exit(1);
-  });
+await main();
```
given we're just awaiting main() at the top level of the module, do we need the main function at all, just make everything top level?  this feels very script-y anyway

le-cong replied:
good idea

le-cong replied:
tried that and making everything top level, it works but there is one advantage in having a 'main' function when it comes to testing.
what happens is that we'll have to do dynamic import in the tests; and before the dynamic import we normally setup nocking and environment variables. Jest has its own module catching mechanism, which causes some of the tests start failing. one trick is to do `jest.resetModules()` everywhere.
it seems easier and less surprise if we call the main function instead.  
what do you think? should we make the change, or keep the main?

carlansley replied:
if we're await()ing main, isn't it effectively having everything top-level anyway?  my thought is we move main() into a separate module, and all the index.ts does is import and await it.  that way tests can do their thing without side effects going on at the module level.

carlansley reviewed (approved):
lftm, the only issue i can think of is would this work with packages that export weird things (e.g. *-config repos).  e.g. jest-config doesn't export anything at all (maybe it should?).  i suppose we can cross that bridge when we come to it.
